File: cpp/bc2026/src/bestsofar/Utils/Vision.cpp
#include "Utils/Vision.hpp"

#include <cmath>

#include "Utils/AdditionalClasses/Thrown.hpp"
#include "Utils/Globals.hpp"

namespace Utils {

int64_t Vision::hasSeen[60] = {};
Vision::IsWallBits Vision::isWall;
Vision::IsDirtBits Vision::isDirt;
int64_t Vision::hasCheeseMine[60] = {};
int64_t Vision::isReachable[60] = {};
Vision::IsWallBits Vision::isDangerous;

vector<RobotInfo> Vision::enemies;
vector<MapLocation> Vision::walls;
int Vision::numWalls = 0;
vector<MapLocation> Vision::mines;
int Vision::numMines = 0;

// mirrors battlecode.common.UnitType's per-type constants
int Vision::visionConeRadiusSquared(UnitType t) {
    switch (t) {
        case UnitType::BABY_RAT: return 20;
        case UnitType::RAT_KING: return 25;
        case UnitType::CAT: return 17;
    }
    return 0;
}

int Vision::visionConeAngle(UnitType t) {
    switch (t) {
        case UnitType::BABY_RAT: return 90;
        case UnitType::RAT_KING: return 360;
        case UnitType::CAT: return 180;
    }
    return 0;
}

// mirrors MapLocation.isWithinDistanceSquared(location, distanceSquared,
// facingDir, theta) (useBottomLeft == false)
bool Vision::isWithinDistanceSquaredCone(MapLocation self, MapLocation location,
                                         int distanceSquared, Direction facingDir, double theta) {
    if (self == location) {
        return true;
    }

    double adjustment = 1e-3;

    bool isValidDistance = self.distanceSquaredTo(location) <= distanceSquared;

    double dx = location.x - self.x;
    double dy = location.y - self.y;

    bool isValidAngle;
    if (facingDir == Direction::all[Direction::CENTER]) {
        isValidAngle = true;
    } else {
        double cosSim = (facingDir.dx * dx + facingDir.dy * dy) /
                        (std::sqrt((dx * dx + dy * dy) * (double)(facingDir.dx * facingDir.dx +
                                                                  facingDir.dy * facingDir.dy)));
        double halfAngle = std::abs(std::acos(cosSim)) * 180.0 / M_PI;
        isValidAngle = halfAngle - adjustment <= theta / 2;
    }
    return isValidDistance && isValidAngle;
}

void Vision::lookAround() {
    // System.out.println(Clock.getBytecodeNum());

    for (const MapInfo& mi : rc.senseNearbyMapInfos()) {
        int x = mi.getMapLocation().x, y = mi.getMapLocation().y;

        if (((hasSeen[x] >> y) & 1) == 0 || ((isWall[x] >> y) & 1) == 0) {
            hasSeen[x] |= 1LL << y;
            if (mi.isWall()) isWall[x] |= 1LL << y;
            if (mi.hasCheeseMine()) hasCheeseMine[x] |= 1LL << y;
            if (mi.isDirt()) isDirt[x] |= 1LL << y;
            else isDirt[x] &= ~(1LL << y);
        }
    }

    // System.out.println(Clock.getBytecodeNum());
}

void Vision::turn(Direction dir) {
    if (dir == Direction::all[Direction::CENTER] || dir == rc.getDirection() || cantTurn ||
        !rc.canTurn())
        return;
    rc.turn(dir);
    lookAround();
}

bool Vision::turnLeft = false;

Direction Vision::bestTurn() {
    int sumX = rc.getLocation().x, sumY = rc.getLocation().y;
    for (int i = 0; i < (int)newestSqueaks.size(); i++) {
        sumX += newestSqueaks[i].getSource().x;
        sumY += newestSqueaks[i].getSource().y;
    }
    sumX = (sumX + ((int)newestSqueaks.size() + 1) / 2) / ((int)newestSqueaks.size() + 1);
    sumY = (sumY + ((int)newestSqueaks.size() + 1) / 2) / ((int)newestSqueaks.size() + 1);
    Direction d = rc.getLocation().directionTo(MapLocation(sumX, sumY)).opposite();
    if (d == Direction::all[Direction::CENTER]) {
        turnLeft = !turnLeft;
        if (turnLeft) return rc.getDirection().rotateLeft().rotateLeft();
        return rc.getDirection().rotateRight().rotateRight();
    }
    return d;
}

void Vision::turnJustBecause() {
    if (rc.canTurn() && enemyRobots.size() == 0 && newestSqueaks.size() == 0) turn(bestTurn());
}

void Vision::senseNearbyEnemies() {
    enemies = rc.senseNearbyRobots(visionConeRadiusSquared(rc.getType()), opponentTeam);
}

bool Vision::seenByEnemies() {
    for (const RobotInfo& ri : enemies) {
        if (isWithinDistanceSquaredCone(rc.getLocation(), ri.location,
                                        visionConeRadiusSquared(ri.getType()), ri.getDirection(),
                                        visionConeAngle(ri.getType()))) {
            return true;
        }
    }
    return false;
}

void Vision::digDirt(MapLocation loc) {
    isDirt[loc.x] &= ~(1LL << loc.y);
}

void Vision::placeDirt(MapLocation loc) {
    isDirt[loc.x] |= 1LL << loc.y;
}

bool Vision::hasSeenLocation(MapLocation loc) {
    return rc.onTheMap(loc) && ((hasSeen[loc.x] >> loc.y) & 1) > 0;
}

bool Vision::hasSeenLocationKnownToBeOnMap(MapLocation loc) {
    return ((hasSeen[loc.x] >> loc.y) & 1) > 0;
}

bool Vision::sensePassability(MapLocation loc) {
    return ((isWall[loc.x] >> loc.y) & 1) == 0 && ((isDirt[loc.x] >> loc.y) & 1) == 0;
}

// Java `isDirt(MapLocation)` -> Vision::IsDirtBits::operator() in Vision.hpp
// Java `isWall(MapLocation)` -> Vision::IsWallBits::operator() in Vision.hpp

bool Vision::canReach(MapLocation loc) {
    return ((isReachable[loc.x] >> loc.y) & 1) > 0;
}

// Java `isDangerous(MapLocation)` -> Vision::IsWallBits::operator() in Vision.hpp

bool Vision::isPassableWithDig(MapLocation loc) {
    return ((isWall[loc.x] >> loc.y) & 1) == 0;
}

bool Vision::isMine(MapLocation loc) {
    return ((hasCheeseMine[loc.x] >> loc.y) & 1) > 0;
}

void Vision::init() {
}

void Vision::startTurn() {
    if (rc.getType() == UnitType::BABY_RAT)
        lookAround();
    // TODO: optimise by detecting canDig nearby or canMove nearby
    for (const AdditionalClasses::Thrown& throwns : alliesThrown) {
        MapLocation loc = throwns.loc.add(throwns.dir);
        if (hasSeenLocation(loc)) {
            isDangerous[loc.x] |= (1LL << loc.y);
        }
        loc = loc.add(throwns.dir);
        if (hasSeenLocation(loc)) {
            isDangerous[loc.x] |= (1LL << loc.y);
        }
    }
}

void Vision::endTurn() {
    for (const AdditionalClasses::Thrown& throwns : alliesThrown) {
        MapLocation loc = throwns.loc.add(throwns.dir);
        if (hasSeenLocation(loc)) {
            isDangerous[loc.x] &= ~(1LL << loc.y);
        }
        loc = loc.add(throwns.dir);
        if (hasSeenLocation(loc)) {
            isDangerous[loc.x] &= ~(1LL << loc.y);
        }
    }
}

}  // namespace Utils

File: py/bc2026/src/bestsofar/Utils/Vision.py
# Port of bestsofar/Utils/Vision.java. See py/CONVENTIONS.md.
import math

from api import *
from Utils.AdditionalClasses.Thrown import Thrown
from Utils.Globals import Globals


class _Bits(list):
    # The Java Vision has BOTH a `public static long[]` field AND a same-named
    # `public static boolean <name>(MapLocation)` method for isWall and
    # isDangerous. A Python class attribute cannot be both, so these are lists
    # that are also callable: `Vision.isWall[x]` indexes the bitset and
    # `Vision.isWall(loc)` runs the method's exact logic (same trick as the
    # C++ port's IsWallBits functor).
    def __init__(self):
        super().__init__([0] * 60)

    def __call__(self, loc):
        # Java method body: ((<field>[loc.x] >> loc.y) & 1) > 0
        return ((self[loc.x] >> loc.y) & 1) > 0


class _IsDirtBits(_Bits):
    def __call__(self, loc):
        # Java isDirt(MapLocation) reads the hasSeen field plus its own bits
        return ((Vision.hasSeen[loc.x] >> loc.y) & 1) > 0 and ((self[loc.x] >> loc.y) & 1) > 0


class Vision(Globals):

    hasSeen = [0] * 60
    isWall = _Bits()
    isDirt = _IsDirtBits()
    hasCheeseMine = [0] * 60
    isReachable = [0] * 60

    isDangerous = _Bits()

    enemies = None  # RobotInfo[]
    walls = None  # MapLocation[]
    numWalls = 0
    mines = None  # MapLocation[]
    numMines = 0

    turnLeft = False

    @staticmethod
    def lookAround():
        # System.out.println(Clock.getBytecodeNum());

        for mi in Globals.rc.senseNearbyMapInfos():
            x = mi.getMapLocation().x
            y = mi.getMapLocation().y

            if ((Vision.hasSeen[x] >> y) & 1) == 0 or ((Vision.isWall[x] >> y) & 1) == 0:
                Vision.hasSeen[x] |= 1 << y
                if mi.isWall():
                    Vision.isWall[x] |= 1 << y
                if mi.hasCheeseMine():
                    Vision.hasCheeseMine[x] |= 1 << y
                if mi.isDirt():
                    Vision.isDirt[x] |= 1 << y
                else:
                    Vision.isDirt[x] &= ~(1 << y)

        # System.out.println(Clock.getBytecodeNum());

    @staticmethod
    def turn(dir):
        if dir == Direction.CENTER or dir == Globals.rc.getDirection() or Globals.cantTurn or not Globals.rc.canTurn():
            return
        Globals.rc.turn(dir)
        Vision.lookAround()

    @staticmethod
    def bestTurn():
        sumX = Globals.rc.getLocation().x
        sumY = Globals.rc.getLocation().y
        for i in range(len(Globals.newestSqueaks)):
            sumX += Globals.newestSqueaks[i].getSource().x
            sumY += Globals.newestSqueaks[i].getSource().y
        # all operands non-negative: Java int division
        sumX = (sumX + (len(Globals.newestSqueaks) + 1) // 2) // (len(Globals.newestSqueaks) + 1)
        sumY = (sumY + (len(Globals.newestSqueaks) + 1) // 2) // (len(Globals.newestSqueaks) + 1)
        d = Globals.rc.getLocation().directionTo(MapLocation(sumX, sumY)).opposite()
        if d == Direction.CENTER:
            Vision.turnLeft = not Vision.turnLeft
            if Vision.turnLeft:
                return Globals.rc.getDirection().rotateLeft().rotateLeft()
            return Globals.rc.getDirection().rotateRight().rotateRight()
        return d

    @staticmethod
    def turnJustBecause():
        if Globals.rc.canTurn() and len(Globals.enemyRobots) == 0 and len(Globals.newestSqueaks) == 0:
            Vision.turn(Vision.bestTurn())

    @staticmethod
    def senseNearbyEnemies():
        # Java: rc.getType().visionConeRadiusSquared
        Vision.enemies = Globals.rc.senseNearbyRobots(Vision.visionConeRadiusSquared(Globals.rc.getType()), Globals.opponentTeam)

    @staticmethod
    def seenByEnemies():
        for ri in Vision.enemies:
            # Java: rc.getLocation().isWithinDistanceSquared(ri.location,
            #   ri.getType().visionConeRadiusSquared, ri.getDirection(),
            #   ri.getType().visionConeAngle)
            if Vision.isWithinDistanceSquaredCone(Globals.rc.getLocation(), ri.location,
                                                  Vision.visionConeRadiusSquared(ri.getType()),
                                                  ri.getDirection(),
                                                  Vision.visionConeAngle(ri.getType())):
                return True
        return False

    @staticmethod
    def digDirt(loc):
        Vision.isDirt[loc.x] &= ~(1 << loc.y)

    @staticmethod
    def placeDirt(loc):
        Vision.isDirt[loc.x] |= 1 << loc.y

    @staticmethod
    def hasSeenLocation(loc):
        return Globals.rc.onTheMap(loc) and ((Vision.hasSeen[loc.x] >> loc.y) & 1) > 0

    @staticmethod
    def hasSeenLocationKnownToBeOnMap(loc):
        return ((Vision.hasSeen[loc.x] >> loc.y) & 1) > 0

    @staticmethod
    def sensePassability(loc):
        return ((Vision.isWall[loc.x] >> loc.y) & 1) == 0 and ((Vision.isDirt[loc.x] >> loc.y) & 1) == 0

    # Java: isDirt(MapLocation), isWall(MapLocation), isDangerous(MapLocation)
    # live on the bitset objects' __call__ (they share their field's name).

    @staticmethod
    def canReach(loc):
        return ((Vision.isReachable[loc.x] >> loc.y) & 1) > 0

    @staticmethod
    def isPassableWithDig(loc):
        return ((Vision.isWall[loc.x] >> loc.y) & 1) == 0

    @staticmethod
    def isMine(loc):
        return ((Vision.hasCheeseMine[loc.x] >> loc.y) & 1) > 0

    @staticmethod
    def init():
        pass

    @staticmethod
    def startTurn():
        if Globals.rc.getType() == UnitType.BABY_RAT:
            Vision.lookAround()
        # TODO: optimise by detecting canDig nearby or canMove nearby
        for throwns in Globals.alliesThrown:
            loc = throwns.loc.add(throwns.dir)
            if Vision.hasSeenLocation(loc):
                Vision.isDangerous[loc.x] |= (1 << loc.y)
            loc = loc.add(throwns.dir)
            if Vision.hasSeenLocation(loc):
                Vision.isDangerous[loc.x] |= (1 << loc.y)

    @staticmethod
    def endTurn():
        for throwns in Globals.alliesThrown:
            loc = throwns.loc.add(throwns.dir)
            if Vision.hasSeenLocation(loc):
                Vision.isDangerous[loc.x] &= ~(1 << loc.y)
            loc = loc.add(throwns.dir)
            if Vision.hasSeenLocation(loc):
                Vision.isDangerous[loc.x] &= ~(1 << loc.y)

    # api.py's UnitType is a plain int and MapLocation only offers the
    # 2-argument isWithinDistanceSquared, so the engine's UnitType vision-cone
    # constants and the cone overload are mirrored here (values/logic copied
    # verbatim from battlecode.common.UnitType / MapLocation, same as the C++
    # port). See notes: these belong in api.py.
    @staticmethod
    def visionConeRadiusSquared(t):
        if t == UnitType.BABY_RAT:
            return 20
        if t == UnitType.RAT_KING:
            return 25
        if t == UnitType.CAT:
            return 17
        return 0

    @staticmethod
    def visionConeAngle(t):
        if t == UnitType.BABY_RAT:
            return 90
        if t == UnitType.RAT_KING:
            return 360
        if t == UnitType.CAT:
            return 180
        return 0

    # mirrors MapLocation.isWithinDistanceSquared(location, distanceSquared,
    # facingDir, theta) (useBottomLeft == false)
    @staticmethod
    def isWithinDistanceSquaredCone(self_, location, distanceSquared, facingDir, theta):
        if self_ == location:
            return True

        adjustment = 1e-3

        isValidDistance = self_.distanceSquaredTo(location) <= distanceSquared

        dx = location.x - self_.x
        dy = location.y - self_.y

        if facingDir == Direction.CENTER:
            isValidAngle = True
        else:
            cosSim = (facingDir.dx * dx + facingDir.dy * dy) / \
                math.sqrt((dx * dx + dy * dy) * (facingDir.dx * facingDir.dx + facingDir.dy * facingDir.dy))
            halfAngle = abs(math.acos(cosSim)) * 180.0 / math.pi
            isValidAngle = halfAngle - adjustment <= theta / 2
        return isValidDistance and isValidAngle
